### PR TITLE
refactor: improvement of chunk scanner performance

### PR DESCRIPTION
**Pull request by MukjepScarlet**
This is related to #3943

It may need a speed test.

### CODE REVIEW

**Comment by MukjepScarlet** [comment]
I ran 3 different tests in a same save (default world settings). All of them used BlockPos.Mutable

Each task iterates 16 * 16 blocks.

| Technique | Time (per chunk) |
| --- | --- |
| Blocking | ~1500us |
| Coroutines (Dispatchers.Default) | ~500us |
| Virtual Threads | ~500us |

**It's down to a third of the time.**

For individual chunks, task performance is essentially consistent along other axes (X & Z).

You can replace `scope.launch` in code with nothing or `Thread.ofVirtual().start` to test. (Don't forget to use `join`)

File: src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/render/ModuleBlockESP.kt [hunk 1]
```diff
@@ -89,30 +87,27 @@ object ModuleBlockESP : Module("BlockESP", Category.RENDER) {
             var dirty = false
 
             renderEnvironmentForWorld(matrixStack) {
-                synchronized(BlockTracker.trackedBlockMap) {
-                    val trackedBlockMap = BlockTracker.trackedBlockMap
-
-                    dirty = drawInternal(this, trackedBlockMap, colorMode, fullAlpha, drawOutline)
-                }
```
Comment (superblaubeere27) [inline src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/render/ModuleBlockESP.kt:96]: 👍🏻

**Review by superblaubeere27 — COMMENTED** [review]
This PR is generally good, I have no complaints about the code. I am not quite sure if the changes will improve the situation.

This improves the chunk scanner *latency*, the performance is probably untouched or might even feel worse for the end user.

Is the latency really a problem? If chunk scanning takes 1.5ms per chunk and the user joins in a world with 16 chunks view range, the scanning would take 16 chunk^(1/2) x 16 chunk^(1/2) x PI x 1.5ms/chunk = 1200 ms. By doing the same task concurrently with three threads, the chunk scanner takes a third of the time, but takes 3 cores from doing other stuff which might make the lagging worse. I think it is better to make the users wait for a few seconds instead.

@1zuna opinion?

**Comment by MukjepScarlet** [comment]
Scan delay os acceptable because it is running with another thread. I think the main cause of the UI stuck is the use of `synchronized` without swithing thread before. `ArrayBlockingQueue` will blocking all add operation if it is full. Now it runs adding on another thread (through `scope.launch`), so users will feel better.

**Comment by MukjepScarlet** [comment]
Scan delay is acceptable because it is running with another thread. I think the main cause of the UI stuck is the use of `synchronized` without swithing thread before. `ArrayBlockingQueue` will blocking all add operation if it is full. Now it runs adding on another thread (through `scope.launch`), so users will feel better.

We can add a boolean option to the config and let the user decide whether to use concurrent scanning or not.

File: src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/render/ModuleBlockESP.kt (hunk 1 above)
Comment (MukjepScarlet) [reply src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/render/ModuleBlockESP.kt:96]: `trackedBlockMap` is already a ConcurrentHashMap so we don't need to `synchronized` it. Next frame the variation will be rendered.

**Pull request merged**